contracts/Stabilizer.sol

```diff
@@ -0,0 +1,155 @@
+        buyFee = amount;
+    }
+
+    function setSellFee(uint amount) public onlyOperator {
+        require(amount <= MAX_FEE, "amount too high");
+        sellFee = amount;
+    }
+    
+    function setCap(uint amount) public onlyOperator {
+        supplyCap = amount;
+    }
+
+    function setGovernance(address gov_) public onlyGovernance {
+        governance = gov_;
+    }
+
+    function setStrat(IStrat newStrat) public onlyGovernance {
+        if(address(strat) != address(0)) {
+            uint prevTotalValue = strat.calcTotalValue();
+            strat.divest(prevTotalValue);
+        }
+        reserve.transfer(address(newStrat), reserve.balanceOf(address(this)));
+        newStrat.invest();
+        strat = newStrat;
+    }
+
+    function removeStrat() public onlyGovernance {
+        uint prevTotalValue = strat.calcTotalValue();
+        strat.divest(prevTotalValue);
+
+        strat = IStrat(address(0));
+    }
+
+    function takeProfit() public {
+        uint totalReserves = getTotalReserves();
+        if(totalReserves > supply) {
+            uint profit = totalReserves - supply; // underflow prevented by if condition
+            if(address(strat) != address(0)) {
+                uint bal = reserve.balanceOf(address(this));
+                if(bal < profit) {
+                    strat.divest(profit - bal); // underflow prevented by if condition
+                }
+            }
+            reserve.transfer(governance, profit);
+        }
+    }
+
+    function buy(uint amount) public {
+        if(address(strat) != address(0)) {
+            reserve.transferFrom(msg.sender, address(strat), amount);
+            strat.invest();
+        } else {
+            reserve.transferFrom(msg.sender, address(this), amount);
+        }
+
+        if(buyFee > 0) {
+            uint fee = amount.mul(buyFee).div(MAX_FEE);
+            reserve.transferFrom(msg.sender, governance, fee);
+        }
+
+        synth.mint(msg.sender, amount);
+        supply = supply.add(amount);
+        require(supply <= supplyCap, "supply exceeded cap");
+    }
+
+    function sell(uint amount) public {
+        synth.transferFrom(msg.sender, address(this), amount);
+        synth.burn(amount);
+
+        uint reserveBal = reserve.balanceOf(address(this));
+        if(address(strat) != address(0) && reserveBal < amount) {
+            strat.divest(amount - reserveBal); // underflow prevented by if condition
+        }
+
+        uint afterFee;
+        if(sellFee > 0) {
+            uint fee = amount.mul(sellFee).div(MAX_FEE);
+            afterFee = amount.sub(fee);
+            reserve.transfer(governance, fee);
+        } else {
+            afterFee = amount;
+        }
+        
+        reserve.transfer(msg.sender, afterFee);
+        supply = supply.sub(amount);
+    }
+
+    function rescue(ERC20 token) public onlyGovernance {
+        token.transfer(governance, token.balanceOf(address(this)));
```

Comment:
Can we exclude the reserve token here?

Reply:
This function is meant to allow gov to rescue `reserve` funds if they are stuck due to a bug. Using it will require a 5-day public governance vote so it will leave plenty of time for users to react.

Reply:
Actually I think I'm gonna add that check to minimize trust assumptions. This contract is simple enough for stuck funds to be fairly improbable.

Reply:
done